### PR TITLE
perf(sampler): fused top-p/top-k/temp sampler — B=1 +42% on Qwen 3.6 35B-A3B

### PR DESCRIPTION
**Pull request by raullenchai**
## Summary

Closes the 4.3 ms / token B=1 gap to mlx-vlm under sampling. **Qwen 3.6 35B-A3B 4-bit @ B=1 HTTP: 65 → 92 tok/s (+42%)**, matches mlx-vlm 92.7. Qwen 3.5 4B sanity: 128 → 137 tok/s (+7%, smaller absolute sampler cost).

Root cause (investigation report previously documented + closed via #541): mlx-lm's `make_sampler` chains independent `@mx.compile`'d ops — `apply_top_p` → `apply_top_k` → `categorical_sampling`. For the dominant chat config (temp + top_p + top_k from alias defaults), this costs ~4.3 ms / token split as:

- **~0.9 ms Python** — three closure dispatches per step inside mlx-lm's per-row sampler loop.
- **~3.3 ms GPU** — two `@mx.compile` boundaries break lazy-graph fusion across `apply_top_p → categorical_sampling`. `apply_top_p` also builds a vocab-sized `inverse_indices` array (`mx.put_along_axis + mx.arange`) to undo the sort permutation back into vocab order so `categorical` can sample in vocab space — that scatter is the largest single op in the chain.

## What this PR ships

- **`vllm_mlx/_sampler_fast_path.py`** — vendored single-function fused sampler. Builds the top-p / top-k mask in **sorted** space (no `inverse_indices`), applies temperature scaling and mask to the sorted logits, samples in sorted space, maps back via one `take_along_axis`.
- **`vllm_mlx/scheduler.py`** — wires the fast path into the existing `_get_request_sampler` cache (4-tuple key on `(temp, top_p, min_p, top_k)`). Eligibility:
  - `temperature > 0` (mlx-lm already short-circuits to argmax at `temp == 0`),
  - `min_p == 0` (would need a fourth op the fast path doesn't implement — falls through),
  - at least one of `0 < top_p < 1` or `top_k > 0` is active (otherwise the chain collapses to plain `categorical_sampling` — no benefit to steal),
  - one-time INFO log on first engagement so reviewers can confirm in production logs.
- **`tests/test_sampler_fast_path.py`** — 15 cases covering eligibility window, shape contract, and distributional equivalence to mlx-lm's chain.

## Math equivalence vs mlx-lm

The fused sampler computes the top-p mask on the unscaled probability distribution (matching `apply_top_p` semantics — *not* mlx-vlm's post-temperature variant), applies the top-k mask, then scales the masked logits by `1/T` before `mx.random.categorical`. Identical to `apply_top_p → apply_top_k → categorical_sampling` in:

- the kept set of tokens (top-p computed pre-temperature, top-k by rank),
- the relative weights inside the kept set,
- the expected sample distribution.

Only sample-level determinism with a fixed `mx.random.seed` differs: mlx-lm draws Gumbel noise in vocab order, we draw in sorted order — same seed picks a different token. Within-engine reproducibility is preserved (same input + same seed always picks the same token on this engine), which is what the OpenAI `seed=` field promises.

## Bench

Qwen 3.6 35B-A3B 4-bit, M3 Ultra 256 GB, B=1 HTTP streaming chat completion, `temperature=0.7 top_p=0.95`, `top_k=20` injected by the alias's generation_config cascade.

| Variant | tok/s |
|---|---|
| baseline (mlx-lm chain) | 65.10 |
| **this PR (3 runs)** | **91.66 / 91.74 / 92.38** |
| mlx-vlm 0.6.1 reference | 92.73 |

Qwen 3.5 4B 4-bit sanity: 128 → 137 tok/s (+7%).

## Test plan

- [x] `pytest tests/test_sampler_fast_path.py` — 15/15 pass
- [x] `pytest tests/test_dense_sampler_fastpath.py tests/test_sampling_params_passthrough.py tests/test_alias_recommended_sampling.py` — 38/38 pass (no regression on existing sampler / cache / alias logic)
- [x] `ruff check vllm_mlx/_sampler_fast_path.py vllm_mlx/scheduler.py tests/test_sampler_fast_path.py` clean
- [x] Live HTTP bench on Qwen 3.6 35B-A3B 4-bit confirms +42% throughput vs baseline
- [x] Live HTTP bench on Qwen 3.5 4B 4-bit confirms +7% throughput (no regression on small models)
- [x] `[fused_top_p_sampler] engaged for ...` INFO log confirmed in production logs

## Risk assessment

- Touches the **per-request sampler factory** (`Scheduler._get_request_sampler`) — single hot path, shared by every dense / hybrid request. Falls through to mlx-lm's chain on every off-window combination (greedy, min_p, no top-p+no top-k), so the **fallback path is untouched**.
- The fast sampler is mathematically equivalent to mlx-lm's chain at the distribution level. Within-engine `seed=` reproducibility is preserved. Cross-engine bit-level reproducibility against a vanilla mlx-lm install is not — same seed picks a different token on the fused path. Documented in the module docstring.
- The existing `_install_dense_sampler_fastpath` (PR #521 batched-sampler swap) continues to work: when multiple homogeneous requests share the cached fused sampler (identity-equal via the sampler cache), PR #521's homogeneous-batch detection kicks in and calls the fused sampler **once on the full `[B, V]` matrix** instead of per-row. So the fast path scales with batch size too.

Generated with Claude Code

### CODE REVIEW

**Comment by raullenchai** [comment]
## Cross-model A/B bench

Same harness, same 256-token streaming chat completion at `temp=0.7 top_p=0.95`, B=1 HTTP on M3 Ultra. A/B done in-process via `RAPID_MLX_DISABLE_FUSED_SAMPLER=1` (the env opt-out added in c6d49a5). Three runs per side, median reported.

| Model | Arch | Vocab | Quant | Fast path tok/s | Baseline tok/s | Delta | Fast path engaged? |
|---|---|---|---|---|---|---|---|
| qwen3.6-35b | hybrid + MoE | 151k | 4-bit | **90.85** | 66.41 | **+36.8%** | yes (top_k=20) |
| qwen3.6-27b-8bit | hybrid | 151k | 8-bit | 19.50 | 19.00 | +2.6% | yes (top_k=0) |
| hermes3-8b (Llama 3.1) | dense | 128k | 4-bit | 116.74 | 115.09 | +1.4% | yes (top_k=0) |
| qwen3.5-4b | dense | 151k | 4-bit | ~137 | ~128 | +7% | yes (top_k=0) |
| gemma3-12b | dense | 262k | 4-bit | 62.37 | 62.88 | -0.81% | **no** — cascade falls outside eligibility window |

### Reading the data

- **The big win is concentrated on Qwen 3.6 35B-A3B** — the hybrid MoE model where (a) the active-param forward is cheap so the sampler is a *large fraction* of per-step wall, and (b) `top_k=20` from the alias cascade triples the per-step sampler-chain cost (apply_top_p + apply_top_k + categorical_sampling, three `@mx.compile` boundaries instead of two).
- **Dense models see ~1-3%** because the absolute sampler cost (1-2 ms) is a small fraction of the forward-dominated step on these architectures. The win is real but in the noise band; no regression.
- **No regressions** on any model.
- **gemma3-12b's fast path did not engage** — the alias's `recommended_sampling` + the model's `generation_config.json` defaults push the cascade outside our eligibility window (likely `min_p > 0` lands on the request). Both sides ran the mlx-lm chain identically; the -0.8% is run-to-run noise, not a fast-path artifact. Extending the fast path to cover `min_p` is a small follow-up if Gemma is a load-bearing target — out of scope here.

### Engagement logs (one-time per Scheduler)

```
[fused_top_p_sampler] engaged for temp=0.700 top_p=0.950 top_k=20  ← qwen3.6-35b
[fused_top_p_sampler] engaged for temp=0.700 top_p=0.950 top_k=0   ← qwen3.6-27b-8bit
[fused_top_p_sampler] engaged for temp=0.700 top_p=0.950 top_k=0   ← hermes3-8b
[fused_top_p_sampler] engaged for temp=0.700 top_p=0.950 top_k=0   ← qwen3.5-4b
(no engagement)                                                     ← gemma3-12b
```

### Headline

This PR is targeted at the largest gap (Qwen 3.6 35B-A3B B=1 sampled, +37%) and does not regress anything else. Dense and smaller-vocab models pick up incremental gains; Gemma 3 12B is a no-op until the fast path covers `min_p`.
